extensions/amp-form/0.1/form-dirtiness.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {isDisabled, isFieldDefault, isFieldEmpty} from '../../../src/form';
+
+export const DIRTINESS_INDICATOR_CLASS = 'amp-form-dirty';
+
+export class FormDirtiness {
+  /**
+   * @param {!HTMLFormElement} form
+   */
+  constructor(form) {
+    /** @private @const {!HTMLFormElement} */
+    this.form_ = form;
+
+    /** @private {number} */
+    this.dirtyFieldCount_ = 0;
+
+    /** @private {!Object<string, boolean>} */
+    this.isFieldNameDirty_ = {};
+
+    /** @private {boolean} */
+    this.isSubmitting_ = false;
+
+    this.installEventListeners_();
+  }
+
+  /**
+   * Processes dirtiness state when a form is being submitted.
+   */
+  onSubmitting() {
+    this.isSubmitting_ = true;
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * Processes dirtiness state when the form submission fails.
+   */
+  onSubmitError() {
+    this.isSubmitting_ = false;
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * Processes dirtiness state when the form submission succeeds.
+   */
+  onSubmitSuccess() {
+    this.isSubmitting_ = false;
+    this.clearDirtyFields_();
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * @private
+   */
+  installEventListeners_() {
+    this.form_.addEventListener('input', this.onInput_.bind(this));
+    this.form_.addEventListener('reset', this.onReset_.bind(this));
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  onInput_(event) {
+    const field = dev().assertElement(event.target);
+    this.checkDirtinessAfterUserInteraction_(field);
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * @param {!Event} unusedEvent
+   * @private
+   */
+  onReset_(unusedEvent) {
```

Comment:
Can this `reset` event targeting on only one field rather than the whole form?

Reply:
Always the form, per https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reset_event

> The `reset` event fires when a `<form>` is reset.